test/Test/Toml/BiCode/Property.hs

```diff
@@ -0,0 +1,191 @@
+module Test.Toml.BiCode.Property where
+
+import Control.Applicative ((<|>))
+import Control.Category ((>>>))
+import Data.ByteString (ByteString)
+import Data.Hashable (Hashable)
+import Data.HashSet (HashSet)
+import Data.IntSet (IntSet)
+import Data.List.NonEmpty (NonEmpty)
+import Data.Set (Set)
+import Data.Text (Text)
+import Data.Time (Day, LocalTime, TimeOfDay, ZonedTime, zonedTimeToUTC)
+import GHC.Exts (fromList)
+import Hedgehog (Gen, forAll, tripping)
+import Numeric.Natural (Natural)
+
+import Toml (TomlBiMap, TomlCodec, (.=))
+import Toml.Bi.Code (decode, encode)
+
+import Test.Toml.Gen (PropertyTest, genBool, genDay, genDouble, genHours, genInteger, genLocal,
+                      genText, genZoned, prop)
+
+import qualified Data.ByteString.Lazy as L
+import qualified Data.Text as T
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Toml
+
+test_encodeDecodeProp :: PropertyTest
+test_encodeDecodeProp = prop "decode . encode == id" $ do
+    bigType <- forAll genBigType
+    tripping bigType (encode bigTypeCodec) (decode bigTypeCodec)
+
+data BigType = BigType
+    { btBool          :: Bool
+    , btInteger       :: Integer
+    , btNatural       :: Natural
+    , btInt           :: Int
+    , btWord          :: Word
+    , btDouble        :: Double
+    , btFloat         :: Float
+    , btText          :: Text
+    , btString        :: String
+    , btBS            :: ByteString
+    , btLazyBS        :: L.ByteString
+    , btLocalTime     :: LocalTime
+    , btDay           :: Day
+    , btTimeOfDay     :: TimeOfDay
+    , btArray         :: [Int]
+    , btArraySet      :: Set Word
+    , btArrayIntSet   :: IntSet
+    , btArrayHashSet  :: HashSet Double
+    , btArrayNonEmpty :: NonEmpty Text
+    , btNonEmpty      :: NonEmpty ByteString
+    , btList          :: [Bool]
+    , btNewtype       :: BigTypeNewtype
+    , btSum           :: BigTypeSum
+    , btRecord        :: BigTypeRecord
+    } deriving (Show, Eq)
+
+newtype BigTypeNewtype = BigTypeNewtype ZonedTime
+    deriving (Show)
+
+instance Eq BigTypeNewtype where
+    (BigTypeNewtype a) == (BigTypeNewtype b) = zonedTimeToUTC a == zonedTimeToUTC b
```

Comment:
Very nice idea to move `ZonedTime` inside `newtype`! With this trick you can minimize the size of `Eq` instance and test both `diwrap` and `zonedTime`!